import sys
import os
import threading
import subprocess
import ctypes
import time
import serial
import serial.tools.list_ports
from PyQt5 import QtWidgets, uic
from PyQt5.QtCore import pyqtSignal, QObject
from PyQt5.QtWidgets import QFileDialog
from serial_bridge_class import SerialBridge  # –ò–º–ø–æ—Ä—Ç–∏—Ä–∞–º–µ –Ω–∞—à–∏—è –±—Ä–∏–¥–∂ –∫–ª–∞—Å

class Communicator(QObject):
    log_signal = pyqtSignal(str)
    status_signal = pyqtSignal(str)
    baudrate_signal = pyqtSignal(str)

class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super(MainWindow, self).__init__()

        try:
            uic.loadUi(self.resource_path("Screen01.ui"), self)
        except Exception as e:
            print(f"–ì—Ä–µ—à–∫–∞ –ø—Ä–∏ –∑–∞—Ä–µ–∂–¥–∞–Ω–µ –Ω–∞ UI: {e}")
            sys.exit(1)

        self.comm = Communicator()
        self.comm.log_signal.connect(self.log_to_console)
        self.comm.status_signal.connect(self.update_status)
        self.comm.baudrate_signal.connect(self.update_baudrate)

        self.btnBoot.clicked.connect(self.boot_controller_mode3)
        self.btnStartFDT.clicked.connect(self.start_fdt)

        self.real_port = None
        self.virt_port = None
        self.bridge = None

        self.refresh_ports()

        if not self.is_admin():
            self.comm.log_signal.emit("–°—ä–∑–¥–∞–≤–∞–Ω–µ—Ç–æ –Ω–∞ –≤–∏—Ä—Ç—É–∞–ª–Ω–∏ –ø–æ—Ä—Ç–æ–≤–µ –∏–∑–∏—Å–∫–≤–∞ –∞–¥–º–∏–Ω–∏—Å—Ç—Ä–∞—Ç–æ—Ä—Å–∫–∏ –ø—Ä–∞–≤–∞.")
            self.comm.log_signal.emit("–ú–æ–ª—è, —Å—ä–∑–¥–∞–π—Ç–µ –≤–∏—Ä—Ç—É–∞–ª–Ω–∏ COM –ø–æ—Ä—Ç–æ–≤–µ —Ä—ä—á–Ω–æ —Å –ø–æ–º–æ—â—Ç–∞ –Ω–∞ com0com.")
            self.comm.log_signal.emit("–°–∫—Ä–∏–ø—Ç—ä—Ç —â–µ –≥–∏ —Ä–∞–∑–ø–æ–∑–Ω–∞–µ –∞–≤—Ç–æ–º–∞—Ç–∏—á–Ω–æ –ø—Ä–∏ —Å–ª–µ–¥–≤–∞—â–æ—Ç–æ —Å—Ç–∞—Ä—Ç–∏—Ä–∞–Ω–µ.")

    def resource_path(self, relative_path):
        """–ü–æ–∑–≤–æ–ª—è–≤–∞ —Ä–∞–±–æ—Ç–∞ —Å—ä—Å —Å–∫—Ä–∏–ø—Ç–∞ –∫–∞—Ç–æ .exe —Å PyInstaller"""
        try:
            base_path = sys._MEIPASS
        except Exception:
            base_path = os.path.abspath(".")
        return os.path.join(base_path, relative_path)

    def is_admin(self):
        try:
            return ctypes.windll.shell32.IsUserAnAdmin()
        except:
            return False

    def refresh_ports(self):
        self.comboRealPort.clear()
        self.comboVirtPort.clear()

        real_ports, virtual_ports = self.get_serial_ports()
        if not real_ports:
            self.comm.log_signal.emit("‚ö† –ù–µ —Å–∞ –Ω–∞–º–µ—Ä–µ–Ω–∏ —Ä–µ–∞–ª–Ω–∏ —Å–µ—Ä–∏–π–Ω–∏ –ø–æ—Ä—Ç–æ–≤–µ!")
        else:
            self.comboRealPort.addItems(real_ports)

        if not virtual_ports:
            self.comm.log_signal.emit("‚ö† –ù–µ —Å–∞ –Ω–∞–º–µ—Ä–µ–Ω–∏ –≤–∏—Ä—Ç—É–∞–ª–Ω–∏ —Å–µ—Ä–∏–π–Ω–∏ –ø–æ—Ä—Ç–æ–≤–µ!")
        else:
            self.comboVirtPort.addItems(virtual_ports)

        self.activate_selected_ports()
        self.start_bridge()

    def get_serial_ports(self):
        ports = serial.tools.list_ports.comports()
        real_ports = []
        virtual_ports = []

        for port in ports:
            description = getattr(port, 'description', '')
            if 'com0com' in description.lower() or 'virtual' in description.lower():
                virtual_ports.append(port.device)
            else:
                real_ports.append(port.device)

        return real_ports, virtual_ports

    def activate_selected_ports(self):
        real_name = self.comboRealPort.currentText()
        virt_name = self.comboVirtPort.currentText()

        if self.real_port and self.real_port.is_open:
            self.real_port.close()
        if self.virt_port and self.virt_port.is_open:
            self.virt_port.close()

        try:
            self.real_port = serial.Serial(real_name, baudrate=9600, timeout=1)
            self.comm.log_signal.emit(f"‚úÖ –£—Å–ø–µ—à–Ω–æ –æ—Ç–≤–æ—Ä–µ–Ω —Ä–µ–∞–ª–µ–Ω –ø–æ—Ä—Ç: {real_name}")
        except Exception as e:
            self.comm.log_signal.emit(f"‚ùå –ù–µ—É—Å–ø–µ—à–Ω–æ –æ—Ç–≤–∞—Ä—è–Ω–µ –Ω–∞ —Ä–µ–∞–ª–µ–Ω –ø–æ—Ä—Ç: {e}")

        try:
            self.virt_port = serial.Serial(virt_name, baudrate=9600, timeout=1)
            self.comm.log_signal.emit(f"‚úÖ –£—Å–ø–µ—à–Ω–æ –æ—Ç–≤–æ—Ä–µ–Ω –≤–∏—Ä—Ç—É–∞–ª–µ–Ω –ø–æ—Ä—Ç: {virt_name}")
        except Exception as e:
            self.comm.log_signal.emit(f"‚ùå –ù–µ—É—Å–ø–µ—à–Ω–æ –æ—Ç–≤–∞—Ä—è–Ω–µ –Ω–∞ –≤–∏—Ä—Ç—É–∞–ª–µ–Ω –ø–æ—Ä—Ç: {e}")

    def start_bridge(self):
        if self.bridge:
            self.bridge.stop()
        if self.real_port and self.virt_port:
            self.bridge = SerialBridge(self.real_port, self.virt_port, self.comm)

    def start_fdt(self):
        def find_fdt_exe():
            paths = []
            for path in os.environ["PATH"].split(os.pathsep):
                exe = os.path.join(path, "FDT.exe")
                paths.append(exe)
                if os.path.isfile(exe):
                    return exe, paths

            fallback = [
                r"C:\Program Files (x86)\Renesas\FDT4.09\FDT.exe",
                r"C:\Renesas\FDT\FDT.exe"
            ]
            for p in fallback:
                paths.append(p)
                if os.path.isfile(p):
                    return p, paths

            return None, paths

        self.comm.log_signal.emit("üîç –¢—ä—Ä—Å—è FDT.exe.")
        fdt_path, paths = find_fdt_exe()

        if not fdt_path:
            self.comm.log_signal.emit("‚ùå FDT.exe –Ω–µ –µ –Ω–∞–º–µ—Ä–µ–Ω –∞–≤—Ç–æ–º–∞—Ç–∏—á–Ω–æ.")
            for p in paths:
                self.comm.log_signal.emit(f"  - {p}")
            fdt_path, _ = QFileDialog.getOpenFileName(self, "–ò–∑–±–µ—Ä–∏ FDT.exe", "", "FDT (*.exe)")
            if not fdt_path:
                self.comm.log_signal.emit("‚ùå FDT –Ω–µ –µ –∏–∑–±—Ä–∞–Ω.")
                return

        try:
            subprocess.Popen([fdt_path], shell=True)
            self.comm.log_signal.emit(f"‚úÖ FDT —Å—Ç–∞—Ä—Ç–∏—Ä–∞–Ω–æ: {fdt_path}")
        except Exception as e:
            self.comm.log_signal.emit(f"‚ùå –ì—Ä–µ—à–∫–∞ –ø—Ä–∏ —Å—Ç–∞—Ä—Ç–∏—Ä–∞–Ω–µ –Ω–∞ FDT: {e}")

    def boot_controller_mode3(self):
        try:
            if not self.real_port or not self.real_port.is_open:
                self.comm.log_signal.emit("‚ùå –†–µ–∞–ª–Ω–∏—è—Ç –ø–æ—Ä—Ç –Ω–µ –µ –æ—Ç–≤–æ—Ä–µ–Ω.")
                return

            self.real_port.setDTR(False)
            self.comm.log_signal.emit("DTR=HIGH (—Ä–µ–∞–ª–Ω–æ) => –û—Å–≤–æ–±–æ–∂–¥–∞–≤–∞ Tx_Break")
            self.real_port.setRTS(True)
            self.comm.log_signal.emit("RTS=LOW (—Ä–µ–∞–ª–Ω–æ) => RESET –∞–∫—Ç–∏–≤–µ–Ω")
            self.delay_precise_ms(200)

            self.real_port.setDTR(True)
            self.comm.log_signal.emit("DTR=LOW (—Ä–µ–∞–ª–Ω–æ) => –ó–∞–µ–º–∞ Tx_Break")
            self.real_port.setRTS(False)
            self.comm.log_signal.emit("RTS=HIGH (—Ä–µ–∞–ª–Ω–æ) => RESET –û—Å–≤–æ–±–æ–¥–µ–Ω")
            self.delay_precise_ms(200)

            self.real_port.setRTS(True)
            self.comm.log_signal.emit("RTS=LOW (—Ä–µ–∞–ª–Ω–æ) => RESET –∞–∫—Ç–∏–≤–µ–Ω")
            self.delay_precise_ms(200)

            self.real_port.setRTS(False)
            self.comm.log_signal.emit("RTS=HIGH (—Ä–µ–∞–ª–Ω–æ) => RESET –û—Å–≤–æ–±–æ–¥–µ–Ω")
            self.delay_precise_ms(200)

            self.real_port.setDTR(False)
            self.comm.log_signal.emit("DTR=HIGH (—Ä–µ–∞–ª–Ω–æ) => –û—Å–≤–æ–±–æ–∂–¥–∞–≤–∞ BREAK")
            self.delay_precise_ms(240)

            for _ in range(16):
                self.real_port.reset_input_buffer()
                self.real_port.write(b'\x00')
                self.comm.log_signal.emit("–ò–∑–ø—Ä–∞—Ç–µ–Ω–æ: 00")
                self.delay_precise_ms(50)

            self.real_port.reset_input_buffer()
            self.real_port.write(b'\xB0')
            self.delay_precise_ms(5)
            response = self.real_port.read(1)

            if response == b'\xB0':
                self.comm.log_signal.emit("‚úÖ –ö–æ–Ω—Ç—Ä–æ–ª–µ—Ä—ä—Ç —Å—Ç–∞—Ä—Ç–∏—Ä–∞ —É—Å–ø–µ—à–Ω–æ –≤ MODE 3.")
            else:
                self.comm.log_signal.emit(f"‚ùå –ö–æ–Ω—Ç—Ä–æ–ª–µ—Ä—ä—Ç –Ω–µ –≤—ä—Ä–Ω–∞ B0. –ü–æ–ª—É—á–µ–Ω–æ: {response.hex().upper() if response else '–ø—Ä–∞–∑–Ω–æ'}")
                return

            self.comm.log_signal.emit("=== –ó–∞–ø–æ—á–≤–∞ ID –ø—Ä–æ–≤–µ—Ä–∫–∞ ===")
            def check_id(id_bytes):
                self.real_port.reset_input_buffer()
                cmd = bytes([0xF5, 0xDF, 0xFF, 0x00, 0x07] + id_bytes)
                self.real_port.write(cmd)
                self.real_port.flush()
                self.comm.log_signal.emit(f"–ò–∑–ø—Ä–∞—Ç–µ–Ω–æ: {cmd.hex().upper()}")
                self.delay_precise_ms(10)

                self.real_port.write(b'\x50')
                self.comm.log_signal.emit("–ò–∑–ø—Ä–∞—Ç–µ–Ω–æ: 50")
                self.delay_precise_ms(10)

                self.real_port.write(b'\x70')
                self.comm.log_signal.emit("–ò–∑–ø—Ä–∞—Ç–µ–Ω–æ: 70")
                self.delay_precise_ms(30)

                available = self.real_port.in_waiting
                resp = self.real_port.read(available)
                self.comm.log_signal.emit(f"–ü–æ–ª—É—á–µ–Ω —Å—Ç–∞—Ç—É—Å: {resp.hex().upper()} (–¥—ä–ª–∂–∏–Ω–∞ {len(resp)})")

                try:
                    i = resp.index(b'\x50')
                    if len(resp) >= i + 4:
                        sdr1 = resp[i + 2]
                        sdr2 = resp[i + 3]
                        if sdr1 == 0x80 and sdr2 == 0x0C:
                            self.comm.log_signal.emit("‚úÖ ID –ø–æ—Ç–≤—ä—Ä–¥–µ–Ω —É—Å–ø–µ—à–Ω–æ (SDR10=1, SDR11=1)")
                            return True
                        else:
                            self.comm.log_signal.emit(f"‚ùå ID –Ω–µ –µ –ø–æ—Ç–≤—ä—Ä–¥–µ–Ω. –ü–æ–ª—É—á–µ–Ω–æ: {sdr1:02X} {sdr2:02X}, –û—á–∞–∫–≤–∞–Ω–æ: 80 0C")
                    else:
                        self.comm.log_signal.emit("‚ùå –°—Ç–∞—Ç—É—Å –æ—Ç–≥–æ–≤–æ—Ä—ä—Ç –Ω–µ —Å—ä–¥—ä—Ä–∂–∞ –¥–æ—Å—Ç–∞—Ç—ä—á–Ω–æ –±–∞–π—Ç–æ–≤–µ —Å–ª–µ–¥ 50")
                except ValueError:
                    self.comm.log_signal.emit("‚ùå –ù–µ –µ –æ—Ç–∫—Ä–∏—Ç–∞ –∫–æ–º–∞–Ω–¥–∞ 50 –≤ –æ—Ç–≥–æ–≤–æ—Ä–∞")

                return False

            if check_id([0x00]*7) or check_id([0xFF]*7):
                self.comm.log_signal.emit("ID –ø—Ä–æ–≤–µ—Ä–∫–∞—Ç–∞ –ø—Ä–µ–º–∏–Ω–∞ —É—Å–ø–µ—à–Ω–æ.")
            else:
                self.comm.log_signal.emit("‚ùå ID –ø—Ä–æ–≤–µ—Ä–∫–∞ –Ω–µ—É—Å–ø–µ—à–Ω–∞.")

        except Exception as e:
            self.comm.log_signal.emit(f"–ì—Ä–µ—à–∫–∞ –ø—Ä–∏ –±—É—Ç–≤–∞–Ω–µ: {e}")

    def delay_precise_ms(self, ms):
        start = time.perf_counter()
        while (time.perf_counter() - start) < (ms / 1000.0):
            pass

    def log_to_console(self, text):
        self.console.append(text)

    def update_status(self, status):
        self.labelStatus.setText(f"–°—Ç–∞—Ç—É—Å: {status}")

    def update_baudrate(self, baudrate):
        self.labelBaudrate.setText(f"–°–∫–æ—Ä–æ—Å—Ç: {baudrate} bps")

if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())
